R-package/R/xgb.DMatrix.R

```diff
@@ -195,10 +299,491 @@ xgb.DMatrix <- function(
+#' (same as the objects returned by calling \link{xgb.QuantileDMatrix}).
+#' \item If passing `as_quantile_dmatrix=FALSE`, will have additional subclass 'xgb.ExternalDMatrix'.
+#' }
+#' @seealso \link{xgb.DataIter}, \link{xgb.ProxyDMatrix}, \link{xgb.QuantileDMatrix}
+#' @examples
+#' library(xgboost)
+#' data(mtcars)
+#'
+#' # this custom environment will be passed to the iterator
+#' # functions at each call. It's up to the user to keep
+#' # track of the iteration number in this environment.
+#' iterator_env <- as.environment(
+#'   list(
+#'     iter = 0,
+#'     x = mtcars[, -1],
+#'     y = mtcars[, 1]
+#'   )
+#' )
+#'
+#' # Data is passed in two batches.
+#' # In this example, batches are obtained by subsetting the 'x' variable.
+#' # This is not advantageous to do, since the data is already loaded in memory
+#' # and can be passed in full in one go, but there can be situations in which
+#' # only a subset of the data will fit in the computer's memory, and it can
+#' # be loaded in batches that are accessed one-at-a-time only.
+#' # The 'proxy_handle' is supplied externally when this function is called,
+#' # and it must be passed to 'xgb.ProxyDMatrix' inside of the call.
+#' iterator_next <- function(iterator_env, proxy_handle) {
+#'   curr_iter <- iterator_env[["iter"]]
+#'   if (curr_iter >= 2) {
+#'     # there are only two batches, so this signals end of the stream
+#'     return(0)
+#'   }
+#'
+#'   if (curr_iter == 0) {
+#'     x_batch <- iterator_env[["x"]][1:16, ]
+#'     y_batch <- iterator_env[["y"]][1:16]
+#'   } else {
+#'     x_batch <- iterator_env[["x"]][17:32, ]
+#'     y_batch <- iterator_env[["y"]][17:32]
+#'   }
+#'
+#'   # Function 'xgb.ProxyDMatrix' must be called manually
+#'   # at each batch with all the appropriate attributes,
+#'   # such as feature names and feature types.
+#'   xgb.ProxyDMatrix(
+#'     proxy_handle = proxy_handle,
+#'     data = x_batch,
+#'     label = y_batch
+#'   )
+#'
+#'   iterator_env[["iter"]] <- curr_iter + 1
+#'   # this signals that the stream continues
+#'   return(iterator_env[["iter"]])
+#' }
+#'
+#' # This moves the iterator back to its beginning
+#' iterator_reset <- function(iterator_env) {
+#'   iterator_env[["iter"]] <- 0
+#' }
+#'
+#' data_iterator <- xgb.DataIter(
+#'   env = iterator_env,
+#'   f_next = iterator_next,
+#'   f_reset = iterator_reset
+#' )
+#' cache_prefix <- tempdir()
+#'
+#' # DMatrix will be constructed from the iterator's batches
+#' dm <- xgb.ExternalDMatrix(data_iterator, cache_prefix, nthread = 1)
+#'
+#' # After construction, can be used as a regular DMatrix
+#' params <- list(nthread = 1, objective = "reg:squarederror")
+#' model <- xgb.train(data = dm, nrounds = 2, params = params)
+#'
+#' # Predictions can also be called on it, and should be the same
+#' # as if the data were passed differently.
+#' pred_dm <- predict(model, dm)
+#' pred_mat <- predict(model, as.matrix(mtcars[, -1]))
+#' @export
+xgb.ExternalDMatrix <- function(
```

Comment:
I'm not sure whether should make a different class at the interface level for external memory. Personally, I would prefer simply having DM and QDM to accept an iterator to avoid having confusion.